### PR TITLE
Redis transport

### PR DESCRIPTION
**Pull request by portertech**
Depends on https://github.com/portertech/em-redis/pull/4

### CODE REVIEW

**Comment by portertech** [comment]
Sensu specs are passing, using the Redis transport :+1: 

```
portertech@d2o:~/projects/sensu/sensu (feature/redis-transport *)$ be rake
/home/portertech/.rubies/2.1.2/bin/ruby -I/home/portertech/.gem/ruby/2.1.2/gems/rspec-core-3.0.4/lib:/home/portertech/.gem/ruby/2.1.2/gems/rspec-support-3.0.4/lib -S /home/portertech/.gem/ruby/2.1.2/gems/rspec-core-3.0.4/exe/rspec ./spec/api/process_spec.rb ./spec/cli_spec.rb ./spec/client/process_spec.rb ./spec/client/socket_spec.rb ./spec/daemon_spec.rb ./spec/server/filter_spec.rb ./spec/server/handle_spec.rb ./spec/server/mutate_spec.rb ./spec/server/process_spec.rb ./spec/utilities_spec.rb
{"timestamp":"2015-05-18T22:51:44.824341-0700","level":"fatal","message":"transport connection error","error":"connection closed"}
...........................................................................................................................

Finished in 1 minute 10.06 seconds (files took 0.1808 seconds to load)
123 examples, 0 failures
```

**Comment by portertech** [comment]
Ready for review :+1:

**Comment by portertech** [comment]
It is possible for JIT Redis connection creation to trigger a reconnect, as `connected?()` may return `false`. Also, the connection monitor is currently enforcing a 3 second timeout :-1:

File: lib/sensu/transport/redis.rb [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+
+      # Publish a message to the Redis transport. The transport pipe
+      # type determines the method of sending messages to consumers
+      # using Redis, either using PubSub or a list. The appropriate
+      # publish method is call for the pipe type given. The Redis
+      # transport ignores publish options.
+      #
+      # @param type [Symbol] the transport pipe type, possible values
+      #   are: :direct and :fanout.
+      # @param pipe [String] the transport pipe name.
```
Comment (grepory) [inline lib/sensu/transport/redis.rb]: Nit: pipe is semantically confusing here.

Comment (portertech) [inline lib/sensu/transport/redis.rb]: "pipe" is the Sensu transport abstraction name.

Comment (grepory) [inline lib/sensu/transport/redis.rb]: That's fair. I does represent an IO target across implementations. Just had to think about it for a minute. :)

File: lib/sensu/transport/redis.rb [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+      #
+      # http://redis.io/topics/pubsub
+      #
+      # @param pipe [String] the transport pipe name.
+      # @param message [String] the message to be published to the transport.
+      # @yield [info] passes publish info to an optional callback/block.
+      # @yieldparam info [Hash] contains publish information.
+      # @yieldparam subscribers [String] current subscriber count.
+      def pubsub_publish(pipe, message, &callback)
+        channel = redis_key("channel", pipe)
```
Comment (grepory) [inline lib/sensu/transport/redis.rb]: If multiple Sensu clusters are backed by the same Elasticache cluster, then this could be potentially dangerous.

Comment (portertech) [inline lib/sensu/transport/redis.rb]: Using a different "db"?

Comment (portertech) [inline lib/sensu/transport/redis.rb]: The Sensu Redis configuration would have to differ anyways, as the installs would collide on keyspace.

Comment (grepory) [inline lib/sensu/transport/redis.rb]: pubsub ignores db. A publish to db 1 will be seen by a subscriber in db 10.

The lists can be managed by using different DBs, but that is then up to configuration management.

Comment (portertech) [inline lib/sensu/transport/redis.rb]: Oh snap! Missed that! Let's add the Redis DB to the keyspace :+1:

**Comment by portertech** [comment]
@grepory :boom:

File: lib/sensu/transport/redis.rb [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+      # `@before_reconnect` transport callback is called.
+      #
+      # @param force [Boolean] the reconnect.
+      def reconnect(force=false)
+        @before_reconnect.call unless @reconnecting
+        unless @reconnecting && !force
+          @reconnecting = true
+          close
+          reset
+          connect
```
Comment (portertech) [inline lib/sensu/transport/redis.rb]: This will reset connection options to an empty hash!

**Comment by portertech** [comment]
We're ready for this to make it into a beta build :+1:

**Pull request merged**